### PR TITLE
Align long-only options in OptionParser help

### PR DESCRIPTION
**Pull request by kojix2**
- Indent long-only options by the short option column
- Keep description alignment unchanged
- Treat empty short flags as long-only options

---

Hello.

While working on YouPlot2, a command-line tool, I noticed that Crystal's `OptionParser` does not indent long-only options in help messages. When options with short options and long-only options are mixed, the option names look misaligned.

Ruby's `OptionParser` indents long-only options by the width of the short option column. YouPlot is written in Ruby, so I was used to this style. I think the same formatting is also natural for Crystal.

I also considered a more compatible approach: adding this indentation only when the same parser has options with both short and long forms.

However, that would make the implementation more complex, with only a small benefit. So this PR uses a simpler approach: always indent long-only options by the short option column, like Ruby does.

This change only affects help message formatting. It does not change how `OptionParser` parses arguments.

The code was developed through several rounds of trial and correction with Copilot and Codex. The final submitted code is mostly based on changes produced by Codex-5.5.

Thank you.

---

YouPlot2 - Crystal

```
Usage: uplot barplot [options] <in.tsv>

Options for barplot:

    -O, --pass              pass input to stdout for pipeline use
    --pass FILE             pass input to FILE for pipeline use
    -o, --output            write plot to stdout
    --output FILE           write plot to FILE (default: stderr)
    -d, --delimiter DELIM   field delimiter (default: TAB)
    -H, --headers           input has a header row
    -T, --transpose         transpose axes of input data
    -t, --title STR         plot title
    --xlabel STR            x-axis label
    --ylabel STR            y-axis label
    -w, --width INT         number of characters per row
    -h, --height INT        number of rows
    -b, --border STR        bounding box style
    -m, --margin INT        spaces to the left of the plot
    --padding INT           spaces left and right of the plot
    -c, --color VAL         drawing color
    --labels                show labels (default)
    --no-labels             hide labels
    -p, --progress          progressive mode [experimental]
    --debug                 print preprocessed data
    --help                  print help
    --version               print version
    --symbol STR            character for bar plots
    --fmt STR               xy (default) or yx
    --xscale STR            x-axis scaling (identity, log10, ln, log2, sqrt, cbrt)
```

YouPlot - Ruby

```
Usage: YouPlot barplot [options] <in.tsv>

Options for barplot:
        --xscale STR        axis scaling (log, ln, log10, lg, log2, lb)
        --fmt STR           xy : header is like x, y...
                            yx : header is like y, x...
        --symbol STR        character to be used to plot the bars

Common options:
    -O, --pass [FILE]       file to output input data to [stdout]
                            for inserting YouPlot in the middle of Unix pipes
    -o, --output [FILE]     file to output plots to [stdout]
                            If no option is specified, plot will print to stderr
    -d, --delimiter DELIM   use DELIM instead of [TAB] for field delimiter
    -H, --headers           specify that the input has header row
    -T, --transpose         transpose the axes of the input data
    -t, --title STR         print string on the top of plot
        --xlabel STR        print string on the bottom of the plot
        --ylabel STR        print string on the far left of the plot
    -w, --width INT         number of characters per row
    -h, --height INT        number of rows
    -b, --border STR        specify the style of the bounding box
                            (solid, corners, barplot)
    -m, --margin INT        number of spaces to the left of the plot
        --padding INT       space of the left and right of the plot
    -c, --color VAL         color of the drawing
        --[no-]labels       hide the labels
    -p, --progress          progressive mode [experimental]
    -C, --color-output      colorize even if writing to a pipe
    -M, --monochrome        no colouring even if writing to a tty
        --encoding STR      specify the input encoding
        --help              print sub-command help menu
        --config FILE       specify a config file
        --debug             print preprocessed data
```

### CODE REVIEW

File: src/option_parser.cr [hunk 1]
```diff
@@ -555,4 +565,10 @@ class OptionParser
       handled
     end
   end
+
+  private def summary_flag?(entry : String) : Bool
+    return false unless entry.starts_with?(summary_indent)
+
+    entry[summary_indent.size..].lstrip.starts_with?('-')
```
Comment (straight-shoota) [inline src/option_parser.cr]: **suggestion:** This could be simplified without any allocations.

```suggestion
    entry.starts_with?(summary_indent) && entry.byte_at?(summary_indent.bytesize).in?('-'.ord, ' '.ord)
```

I'm also wondering if it's really useful extracting this to a helper method.

File: src/option_parser.cr [hunk 2]
```diff
@@ -222,7 +222,11 @@ class OptionParser
     check_starts_with_dash short_flag, "short_flag", allow_empty: true
     check_starts_with_dash long_flag, "long_flag"
 
-    append_flag "#{short_flag}, #{long_flag}", description
+    if short_flag.empty?
+      append_flag long_flag, description
+    else
+      append_flag "#{short_flag}, #{long_flag}", description
+    end
```
Comment (straight-shoota) [inline src/option_parser.cr:230]: **issue:** This looks like an unrelated change. We should extract it to a separate PR.

File: src/option_parser.cr [hunk 3]
```diff
@@ -328,16 +332,22 @@ class OptionParser
   property summary_indent : String = "    "
 
   private def append_flag(flag, description)
+    flag = summary_flag(flag)
```
Comment (straight-shoota) [inline src/option_parser.cr]: **polish:** Extracting this simple helper method doesn't seem to make much sense. IMO the code would be much easier to understand if we inline it.
```suggestion
    # Add indent for long-only options to align with those following a short option
    flag = "    #{flag}" if flag.starts_with?("--")
```

File: src/option_parser.cr [hunk 4]
```diff
@@ -328,16 +332,22 @@ class OptionParser
...
     description_indent = "#{summary_indent}#{" " * summary_width} "
     description = description.gsub("\n", "\n#{description_indent}")
 
-    if flag.size >= summary_width
+    if flag.size > summary_width
```
Comment (straight-shoota) [inline src/option_parser.cr:341]: **issue:** This looks like an unrelated change. We should extract it to a separate PR.

File: src/option_parser.cr (hunk 2 above)
Comment (kojix2) [reply src/option_parser.cr:230]: This is related. `opt.on("", "--foo")` is an existing supported pattern, and it is covered by specs. 
In this case, the empty short flag and comma should be omitted.

File: src/option_parser.cr (hunk 1 above)
Comment (kojix2) [reply src/option_parser.cr]: The revised code is easier to understand, but it's long, so I'd like to break it up into multiple lines.

Comment (kojix2) [reply src/option_parser.cr]: This condition is slightly looser than the previous one.

File: src/option_parser.cr (hunk 4 above)
Comment (kojix2) [reply src/option_parser.cr:341]: I had some confusion about this, but it seems this change was necessary after all.
As it happens, one of the tests failed this time due to a boundary condition. Based on the behavior of Ruby’s OptionParser, this fix is indeed necessary.
However, it is unrelated to this pull request.

Comment (kojix2) [reply src/option_parser.cr:341]: I've created a new pull request https://github.com/crystal-lang/crystal/pull/16915

File: src/option_parser.cr (hunk 1 above)
Comment (straight-shoota) [reply src/option_parser.cr]: Yes, but I think that's acceptable. It would only cause a false positive if the user-provided summary starts with a whitespace, which seems rather untypical.

Happy to split the expression over multiple lines.

**Review by straight-shoota — APPROVED** [review]

File: src/option_parser.cr (hunk 1 above)
Comment (kojix2) [reply src/option_parser.cr]: Thanks for checking.

I agree that the `&&` form is easier to read than the early return, so I changed it here:
https://github.com/crystal-lang/crystal/pull/16914/changes/d2ae9df07b8dd9b51c17645a8adf4d6fa33efcca

But `lstrip` is still needed. For long-only options, `append_flag` adds spaces after `summary_indent` so they align with options that have a short flag. If we don't skip those spaces, `summary_flag?` can't find long-only options, and they are removed from subcommand help.

File: src/option_parser.cr (hunk 2 above)
Comment (kojix2) [reply src/option_parser.cr:230]: Added comment https://github.com/crystal-lang/crystal/pull/16914/changes/011b858ed007af879361fd6b1fd4bf8c363fa80f

**Review by straight-shoota — APPROVED** [review]

**Review by ysbaddaden — APPROVED** [review]

**Comment by kojix2** [comment]
Thanks!
